src/Integration.Vsix/SonarLintNotificationsPackage.cs

```diff
@@ -46,40 +49,84 @@ public sealed class SonarLintNotificationsPackage : Package
+            return System.Threading.Tasks.Task.CompletedTask;
+        }
+
+        private async System.Threading.Tasks.Task Init()
+        {
+            // Working on background thread...
+            Debug.Assert(!ThreadHelper.CheckAccess());
 
-            var sonarqubeService = this.GetMefService<ISonarQubeService>();
-            sonarLintOutput = this.GetMefService<ILogger>();
+            var sonarqubeService = await this.GetMefServiceAsync<ISonarQubeService>();
+            logger = await this.GetMefServiceAsync<ILogger>();
+            logger.WriteLine(Resources.Strings.Notifications_Initializing);
+
+            AddOptionKey(NotificationDataKey);
 
             notifications = new SonarQubeNotificationService(sonarqubeService,
-                new NotificationIndicatorViewModel(), new TimerWrapper { Interval = 60000 }, sonarLintOutput);
+                new NotificationIndicatorViewModel(), new TimerWrapper { Interval = 60000 }, logger);
+
+            activeSolutionBoundTracker = await this.GetMefServiceAsync<IActiveSolutionBoundTracker>();
 
+            // A bound solution might already have completed loading. If so, we need to
+            // trigger the load of the options from the solution file
+            if (activeSolutionBoundTracker.CurrentConfiguration.Mode != SonarLintMode.Standalone)
+            {
+                var solutionPersistence = (IVsSolutionPersistence)await this.GetServiceAsync(typeof(SVsSolutionPersistence));
+                solutionPersistence.LoadPackageUserOpts(this, NotificationDataKey);
+            }
+
+            // Initialising the UI elements has to be on the main thread
+            await JoinableTaskFactory.SwitchToMainThreadAsync();
+            SafePerformOpOnUIThread(() =>
+            {
+                PerformUIInitialisation();
+                logger.WriteLine(Resources.Strings.Notifications_InitializationComplete);
+            });
+        }
+        private void PerformUIInitialisation()
+        {
+            notificationIcon = new NotificationIndicator();
             notificationIcon.DataContext = notifications.Model;
 
-            activeSolutionBoundTracker = this.GetMefService<IActiveSolutionBoundTracker>();
+            // The package is now loaded asynchronously, so a solution might have
+            // finished loading before this package is initialized
+            Refresh(activeSolutionBoundTracker.CurrentConfiguration);
+
+            // Ordering: don't register for solution change notifications
+            // until we've finished setting up everything else
             activeSolutionBoundTracker.SolutionBindingChanged += OnSolutionBindingChanged;
         }
 
         private void OnSolutionBindingChanged(object sender, ActiveSolutionBindingEventArgs binding)
         {
-            if (binding.Configuration.Mode != NewConnectedMode.SonarLintMode.Standalone)
+            SafePerformOpOnUIThread(() => Refresh(binding.Configuration));
```

Comment:
Miscellaneous: I've tried to pull methods called on a UI thread inside a _try-catch_ block so unhandled exceptions won't crash VS.